beard/clustering/blocking.py

```diff
@@ -84,41 +97,54 @@ def _validate(self, X, blocks):
 
         return X, blocks
 
-    def _fit(self, X, blocks):
+    def _fit(self, X, y, blocks):
         """Fit base clustering estimators on X."""
         self.labels_ = -np.ones(len(X), dtype=np.int)
         offset = 0
 
         for b in np.unique(blocks):
-            # Fit on the block
+            # Select data from block
             mask = (blocks == b)
             X_mask = X[mask, :]
 
+            if y is not None:
+                y_mask = y[mask]
+            else:
+                y_mask = None
+
             if self.affinity == "precomputed":
                 X_mask = X_mask[:, mask]
 
-            if self.fit_:
-                cluster = clone(self.base_estimator)
-                cluster.fit(X_mask)
-
+            # Fit a clusterer on the selected data
+            if len(X_mask) == 1:
+                clusterer = _SingleClustering()
+            elif self.fit_:
+                clusterer = clone(self.base_estimator)
             elif self.partial_fit_:
                 if b in self.clusterers_:
-                    cluster = self.clusterers_[b]
+                    clusterer = self.clusterers_[b]
                 else:
-                    cluster = clone(self.base_estimator)
+                    clusterer = clone(self.base_estimator)
 
-                if hasattr(cluster, "partial_fit"):
-                    cluster.partial_fit(X_mask)
-                else:
-                    cluster.fit(X_mask)
+            if self.fit_ or not hasattr(clusterer, "partial_fit"):
+                try:
+                    clusterer.fit(X_mask, y=y_mask)
+                except:
+                    clusterer.fit(X_mask)
+            elif self.partial_fit_:
+                try:
+                    clusterer.partial_fit(X_mask, y=y_mask)
+                except:
+                    clusterer.partial_fit(X_mask)
```

Comment:
Isn't it too ignoring?

If there is an exception in `fit` method which is not caused by the lack of `y_mask` we will see an error (or not) from the second method. If you catch only the proper exception (not sure which one will happen here), we can omit the problem.

The same applies two lines below.